File: withPython/278-first-bad-version/278-first-bad-version.py
# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:

class Solution:
    def firstBadVersion(self, n: int) -> int:
        lt = 1
        rt = n
        def binarySearch(lt, rt):
            mid = (lt + rt) // 2
            mid_center = isBadVersion(mid) 
            mid_left = isBadVersion(mid - 1)
            mid_right = isBadVersion(mid + 1)
            
            if mid_left == False and mid_center == True:
                return mid
            if mid_center == False and mid_right == True:
                return mid + 1
            
            if mid_left == True and mid_center == True:
                return binarySearch(lt, mid - 1)
            
            if mid_center == False and mid_right == False:
                return binarySearch(mid + 1, rt)
            
        return binarySearch(lt, rt)
            
                
        
        

File: 278-first-bad-version/278-first-bad-version.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        int left = 1;
        int right = n;
        int mid = ((right - left)  / 2) + left;
        
        for (int i = 0; i < n; i++) {
            if (!isBadVersion(mid)) {
                left = mid + 1;
                mid = ((right - left)  / 2) + left;
                continue;
            } else {
                if(mid == 1) {
                    return mid;
                }
                if(!isBadVersion(mid - 1)) {
                    return mid;                    
                } 
                if(isBadVersion(mid-1)) {
                    right = mid - 1;
                    mid = ((right - left)  / 2) + left;
                    continue;
                }                
            }
        }
        return -1;
    }
}